opm/autodiff/FullyImplicitBlackoilSolver.hpp

```diff
@@ -194,6 +194,12 @@ namespace Opm {
         std::vector<int>         primalVariable_;
 
         // Private methods.
+
+        // return true if wells are available
+        bool wellsActive() const { return wells_ ? wells_->number_of_wells > 0 : false ; }
+        // return wells object
+        const Wells& wells () const { assert( wells_ ); return *wells_; }
```

Comment:
> Are you sure?

Yes

>
```C++
void assert (int expression);
```

That's wrong.  `Assert` is a macro and therefore doesn't have a formal prototype/declaration.  The `int` part really means that you must pass an `int` (or, in C++, a `bool`).  There is no automatic type conversion going on.

> If this really is illegal, I fear I have done a few `assert(ptr)` errors over the years

It's a common mistake.  I've done it myself on numerous occasions.  Furthermore, the situation is even worse in C than in C++.  Of course, the typical outcome is that the statement behaves exactly in way the programmer expects...

> I meant an example where `assert( wells_ )` would not work.

Not off the top of my head, no.  I suppose a machine with a segmented architecture such as the (defunct) Data General Eclipse would be a candidate.  That machine had different register sets for pointers and integers so you would get more or less a random integer in the `assert` expression unless you explicitly converted the value to an integer.